### PR TITLE
Test and fix AsyncLoader

### LINKED ISSUE #4517: AsyncLoader doesn't cancel on dispose
I'm looking at implementing something similar to `AsyncLoader` that works with `Func<Task>` for use in #4501. I thought a good way to start would be to write some unit tests around `AsyncLoader`.

In the discussion around #4501, @jbialobr [mentions](https://github.com/gitextensions/gitextensions/pull/4501#pullrequestreview-97726773):

> We have a class AsyncLoader that helps us dealing with cancellation and disposed controls. It cancels the previous task if a new is scheduled. It implements IDisposable, so compiler warns if it is not disposed. It cancels the current task on being disposed. 

One of my unit tests attempts to verify that `AsyncLoader.Dispose` prevents the `onLoaded` callback from being invoked. That test fails, and looking at the implementation the code calls `Dispose` on the `CancellationTokenSource` but not `Cancel`.

I'll make a new PR that addresses this issue soon, along with any other issues unit testing turn up. Please shout if there's anything I might be missing.

### PR DESCRIPTION
**Pull request by drewnoakes**
This PR fixes #4517.

Following #4501 I started looking at the behaviour of `AsyncLoader`
and adding some tests. This flagged up a mismatch between the intention
and behaviour of the class. My longer term goal is to implement a similar class that supports `Func<Task>` for use in #4501.

This PR adds thorough tests to `AsyncLoader`, and fixes the bug around
cancellation during disposal.

It also allows more than one thread to specify `DefaultContinuationTaskScheduler`
via the use of `ThreadLocal`.

Added more documentation as my understanding grew.

Also prevents use-after-dispose via `ObjectDisposedException`.

Has been tested on:
 - GIT 2.16
 - Windows 10

### CODE REVIEW

**Comment by drewnoakes** [comment]
Will look at the failing test tomorrow.

**Comment by drewnoakes** [comment]
Fixed floppy test.

**Comment by RussKie** [comment]
Just to confirm, the current plan is to write tests before a major refactor?

**Comment by sharwell** [comment]
How were you expecting us to review this? When I review a style change, I'm on the watch for any behavior change since behavior changes won't be included in a style change. When I'm on the watch for behavior changes, I'm assuming that any change in the diff is part of the behavior change, and it's very confusing and time consuming when I hit code that appears the same (because you know there's a difference, but you just haven't found it yet). This pull request tries to do both at the same time. 😦

**Comment by RussKie** [comment]
I am finding it difficult too to review the changes. The diff is pretty
destructive and I had left a number of comments which later appeared were
against the existing code....

Could you restructure the PR?

File: GitCommands/AsyncLoader.cs [hunk 1]
```diff
@@ -1,149 +1,199 @@
...
         }
 
-        public event EventHandler<AsyncErrorEventArgs> LoadingError = delegate { };
+        #endregion
+
+        public event EventHandler<AsyncErrorEventArgs> LoadingError;
+
+        [NotNull] private readonly TaskScheduler _continuationTaskScheduler;
+        [CanBeNull] private CancellationTokenSource _cancellationTokenSource;
+        private int _disposed;
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitCommands/AsyncLoader.cs]: Can do. My preference is to separate static from instance members. I will move auto properties too, as they're stateful members.

Looking forward to a system that captures these preferences to reduce this back and forth.

File: GitCommands/AsyncLoader.cs [hunk 2]
```diff
@@ -1,149 +1,199 @@
...
-        {
-            AsyncLoader loader = new AsyncLoader();
-            return loader.Load(doMe, continueWith);
-        }
-
-        public Task Load(Action loadContent, Action onLoaded)
-        { 
-            return Load((token) => loadContent(), onLoaded);
-        }
+        public Task Load(Action loadContent, Action onLoaded) => Load(token => loadContent(), onLoaded);
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitCommands/AsyncLoader.cs]: Ditto, re systematic ruling on such matters of style.

File: GitCommands/AsyncLoader.cs [hunk 3]
```diff
@@ -166,46 +217,43 @@ public Task<T> Load<T>(Func<CancellationToken, T> loadContent, Action<T> onLoade
...
-        protected virtual void Dispose(bool disposing)
+        public void Cancel()
         {
-            if (disposing)
-                if (_cancelledTokenSource != null)
-                    _cancelledTokenSource.Dispose();
+            if (Volatile.Read(ref _disposed) != 0)
+                throw new ObjectDisposedException(nameof(AsyncLoader));
+
+            _cancellationTokenSource?.Cancel();
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitCommands/AsyncLoader.cs]: In the solution:

- 17 in-anger uses of `Cancel`
- 15 in-anger uses of `Dispose`
- 11 `Cancel` calls were followed by `Dispose`
- 1 `Cancel` call (in a `Dispose` method itself) was missing a call to `Dispose`
- The other 5 `Cancel` calls seem legitimate to me

Specific use cases include `DashboardItem.CancelBranchNameLoad` which, as the name suggests, cancels any active load operation, yet might still need the `AsyncLoader` instance later on. The other usages seem to have a similar theme.

I've pushed a change that removes redundant calls to `Cancel` before `Dispose` (7a87549b347beff7a310b78336be1e0db4b99f8b) and to correctly dispose the `AsyncLoader` in one case (353223160cffd0ae44f1ac30507244d9184106b2).

The primary motivation here is, as @jbialobr said earlier, disposing the `AsyncLoader` should prevent it later calling back onto the original context (usually the UI) and experiencing exceptions due to controls being disposed. Before this PR you had to remember to call both (and from the numbers above it's clear that some cases were missed). With this design so long as you dispose the instance you don't have to remember to do anything else for things to work.

File: UnitTests/GitCommandsTests/Helpers/AsyncLoaderTests.cs [hunk 4]
```diff
@@ -0,0 +1,244 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using GitCommands;
+using NUnit.Framework;
+
+namespace GitCommandsTests.Helpers
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply UnitTests/GitCommandsTests/Helpers/AsyncLoaderTests.cs]: Done in 65be56c7e824870f7185dafcecbc1e2eebffce3b

File: GitCommands/AsyncLoader.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitCommands/AsyncLoader.cs]: 407ef0e4e6b4df800de7a5827139101786fa1e05

File: GitCommands/AsyncLoader.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitCommands/AsyncLoader.cs]: 407ef0e4e6b4df800de7a5827139101786fa1e05

File: UnitTests/GitCommandsTests/Helpers/AsyncLoaderTests.cs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            _loader.Dispose();
+        }
+
+        [Test]
+        public async Task RegularCase()
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply UnitTests/GitCommandsTests/Helpers/AsyncLoaderTests.cs]: Done in b90ccd00cbfbc6cf062bae73405c663e1302bad1

**Comment by RussKie** [comment]
Please rebase/fix conflicts and we'll get it in.

**Comment by drewnoakes** [comment]
Rebase done.

**Pull request merged**